src/aiida_quantumespresso/workflows/xspectra/crystal.py

```diff
@@ -0,0 +1,676 @@
+
+        structures_to_process = self.ctx.structures_to_process
+        equivalent_sites_data = self.ctx.equivalent_sites_data
+        abs_atom_marker = self.inputs.abs_atom_marker.value
+
+        for site in structures_to_process:
+            inputs = AttributeDict(self.exposed_inputs(XspectraCoreWorkChain, namespace='core'))
+            structure =structures_to_process[site]
+            inputs.structure = structure
+            abs_element = equivalent_sites_data[site]['symbol']
+
+            if 'core_hole_treatments' in self.inputs:
+                ch_treatments = self.inputs.core_hole_treatments.get_dict()
+                ch_treatment = ch_treatments.get(abs_element, 'full')
+            else:
+                ch_treatment = 'full'
+
+            inputs.metadata.call_link_label = f'{site}_xspectra'
+            inputs.eps_vectors = orm.List(list=self.ctx.eps_vectors)
+            if 'core_wfc_data' in self.inputs:
+                inputs.core_wfc_data = self.inputs.core_wfc_data[abs_element]
+            else:
+                inputs.core_wfc_data = self.ctx[f'upf2plotcore_{abs_element}'].outputs.stdout
+
+            # Get the given settings for the SCF inputs and then overwrite them with the
+            # chosen core-hole approximation, then apply the correct pseudopotential pair
+            scf_params = inputs.scf.pw.parameters.get_dict()
+            ch_inputs = XspectraCoreWorkChain.get_treatment_inputs(treatment=ch_treatment)
+            new_scf_params = recursive_merge(left=scf_params, right=ch_inputs)
+
+            # Set the parameter `xiabs` correctly, now that we know which structure we have
+            new_xs_params = inputs.xs_prod.xspectra.parameters.get_dict()
+            kinds_present = sorted([kind.name for kind in structure.kinds])
+            abs_species_index = kinds_present.index(abs_atom_marker) + 1
+            new_xs_params['INPUT_XSPECTRA']['xiabs'] = abs_species_index
+
+            # Set `starting_magnetization` if we are using an XCH approximation, using the
+            # absorbing species as a reasonable place for the unpaired electron.
+            if ch_treatment == 'xch_smear':
+                new_scf_params['SYSTEM'][f'starting_magnetization({abs_species_index})'] = 1
+
+            core_hole_pseudo = self.inputs.core_hole_pseudos[abs_element]
+            gipaw_pseudo = self.inputs.gipaw_pseudos[abs_element]
+            inputs.scf.pw.pseudos[abs_atom_marker] = core_hole_pseudo
+            inputs.scf.pw.pseudos[abs_element] = gipaw_pseudo
+
+            inputs.scf.pw.parameters = orm.Dict(dict=new_scf_params)
+            inputs.xs_prod.xspectra.parameters = orm.Dict(dict=new_xs_params)
+
+            inputs = prepare_process_inputs(XspectraCoreWorkChain, inputs)
+
+            future = self.submit(XspectraCoreWorkChain, **inputs)
+            self.to_context(xspectra_core_workchains=append_(future))
```

Comment:
Is there any special reason to use `append_` instead of using a `dict`? In line 628, you load the `keys` of `structures_to_process` again, thus you assume the keys are always in the same order. This has been the case since Python 3.6, but better to not rely on this.

Reply:
The argument to use `append_` applies when you don't need the keys. In that case, you would often end up creating keys like `f'key_index_{i}` just to have different keys, but if you just care about the values, then a list is more appropriate. Consider this:
```python
def submit(self):
    for i in range(n):
        node = submit(...)
        self.to_context(f'namespace.node_{i}'=node)

def analyse(self):
    for node in self.ctx.namespace:
        ...
```
versus
```python
def submit(self):
    for i in range(n):
        node = submit(...)
        self.to_context(namespace=append_(node))

def analyse(self):
    for node in self.ctx.namespace:
        ...
```
In the end the difference is small and subjective, but I think the second is cleaner since if you don't need the keys, there is no point in creating them.

Reply:
Thanks for the clarification @sphuber, much appreciated!

On reflection, I think for this WorkChain we should convert it to use keys. In the case of the `CoreWorkChain` the important detail in that case was the polarisation vector used for the `XspectraCalculation`, though this information can be retrieved from the input parameters of the node if we need to know for certain - which is what is done in post-processing automatically.

For the `CrystalWorkChain`, the important detail is which site the core-hole is localised at. This can in principle be obtained from the structure used for the calculation, but would be much more laborious than just assigning the key beforehand - and also more useful for the user if they wish to retrieve the information later on - so I will just convert this to use explicit keys.